fix: Mount URL with spaces

EyitopeIO opened the pull request:
To prevent PR from being too large, the path resolution was kept separate and not merged into one. I thought perhaps this would be a refactoring for later. Looking forward to the review and what to change.

**Tests on Ubuntu 22**
```
httpdirfs -d -f "http://0.0.0.0:9000/SP ACE/G B" foo

httpdirfs -d -f --cache --single-file-mode "http://0.0.0.0:9000/SP ACE/G B/code.deb" foo

httpdirfs -d --cache https://old-releases.ubuntu.com/releases/23.04 foo

httpdirfs -d --single-file-mode https://download.iopsys.eu/iopsys/toolchain/crosstools-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32-Rel1.12-full.tar.bz2 foo

httpdirfs -f https://www.crawler-test.com/urls/url_with_spaces/URL%20with%20spaces foo
```

EyitopeIO commented:
Fix for https://github.com/fangfufu/httpdirfs/issues/86

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,57 @@ char *path_append(const char *path, const char *filename)
 
+char *escape_char(const char *url, const ESCAPE_CHAR c)
+{
+    static char escaped[MAX_PATH_LEN];
```
Have you seen this pattern being used elsewhere in this code base? If so, please flag then up. 

And no, please don't do this. I don't think this is threadsafe. Another thread might come in and call this function again, while this function is already called.

EyitopeIO replied:
https://github.com/fangfufu/httpdirfs/pull/137/commits/bedfdf28afa75040a624ddb7834b28fcdf952fab

fangfufu commented:
> To prevent PR from being too large, the path resolution was kept separate and not merged into one. 

Yes this is a good idea.

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,57 @@ char *path_append(const char *path, const char *filename)
+    }
+
+    escaped[j] = '\0';
+    return escaped;
```
Don't do this.

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,57 @@ char *path_append(const char *path, const char *filename)
+    int j = 0;
+    for (int i = 0; i < len; i++) {
+        if (url[i] == escape_me) {
+            switch (c) {
```
Do we really need a switch statement here?

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,61 @@ char *path_append(const char *path, const char *filename)
+            lprintf(fatal, "URL too long: %s\n", url);
+        }
+        if (url[i] == escape_me) {
+            switch (c) {
```
Do we really need a switch statement here?

fangfufu commented on src/util.h:
```diff
@@ -9,6 +9,13 @@
+/**
+ * \brief URL characters that can be escaped
+ */
+typedef enum  {
```
Why don't we specify the actual list of characters that can be escaped here? This is very indirect.

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,61 @@ char *path_append(const char *path, const char *filename)
+
+    switch (c) {
+        case ESCAPED_CHAR_SPACE:
+            escape_me = ' ';
```
I don't think this is a good idea, the characters that need to be escape is not going to change on the fly. You should define it globally, rather than creating a new type to wrap around it. I don't think we need this many indirection.

EyitopeIOn commented:
@fangfufu Replying all comments as @EyitopeIO 

I would look for examples on how to reduce/remove the indirection and update in a bit. I'm just recalling that I've seen good example in [libnetfilter_conntrack](https://www.netfilter.org/projects/libnetfilter_conntrack/index.html) when updating patches.

fangfufu commented:
> @fangfufu Replying all comments as @EyitopeIO
> 
> I would look for examples on how to reduce/remove the indirection and update in a bit. I'm just recalling that I've seen good example in [libnetfilter_conntrack](https://www.netfilter.org/projects/libnetfilter_conntrack/index.html) when updating patches.

Could you point me to your example please? All I am suggesting is that we don't unnecessarily define new types. I understand you might want to do what you did when there might be multiple characters that need to be escaped, but this is not the case here.

EyitopeIO commented:
The "examples" I thought I'd seen in libnetfilter_conntrack source is unrelated.

I've added https://github.com/fangfufu/httpdirfs/pull/137/commits/eb274b303093c185002046e8d6a22f3b692e2529 to address unresolved issues. I hope I understood them correctly. Please don't hesitate to let me know as I'm willing to adapt and learn.

fangfufu commented:
Please could you squash your 3 commits together into one commit, your 3 commits are all addressing the same issue. It really should be squashed together. 

In future consider using ``git commit --amend`` then ``git push --force``.

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,53 @@ char *path_append(const char *path, const char *filename)
     return str;
 }
 
+char *escape_char(const char *url, const char escape_me)
```
We are only escaping a single space character, please rename this function into ``escape_space`` or something. and remove the ``escape_me`` variable.

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,53 @@ char *path_append(const char *path, const char *filename)
+    int replacement_len = 3;
+
+    const char* replacement;
+    if (escape_me == ' ') {
+        replacement = "%20";
+    } else {
+        lprintf(warning, "Not escaping: '%c' in URL\n", escape_me);
+        return NULL;
+    }
```
This is no longer necessary.

fangfufu commented on src/util.h:
```diff
@@ -17,6 +17,16 @@
  */
 char *path_append(const char *path, const char *filename);
 
+/**
+ * \brief escape a single character in a URL
+ * \details This function escapes a character a URL and returns a pointer to
+ * the string with the escaped character. We don't use curl_easy_escape() on
+ * the entire URL because it would break the URL. For example, 'http://a c'
+ * becomes 'http:%2F%2Fa%20c', escaping more characters than we want.
+ * \note You need to free the char * after use.
+ */
+char *escape_char(const char *s, const char c);
```
Please rewrite the description to escaping a single space.

EyitopeIO commented:
@fangfufu 
Squashed commits to one. https://github.com/fangfufu/httpdirfs/pull/137#issuecomment-2067416751
Updated function name and description.  https://github.com/fangfufu/httpdirfs/pull/137#discussion_r1573074530

fangfufu commented on src/util.c:
```diff
@@ -49,6 +49,47 @@ char *path_append(const char *path, const char *filename)
 
+char *escape_spaces(const char *path)
+{
+    char space = ' ';
```
We don't need this variable. Remove please.

EyitopeIO replied:
Commit updated.

fangfufu commented:
Actually I have an important question, why don't we use ``curl_easy_escape``? Why are you writing a function which escape a single space character/

EyitopeIO commented:
@fangfufu Using `curl_easy_escape` unfortunately does not allow us to specify the exact character we want to escape. It escapes everything in the provided string.

Ideally, I think you'd want to `curl_easy_escape(url, len, ' ')`, but instead it escapes everything that can be escaped, including the forward slashes in the URL; thus breaking the URL.

so `http://foo/b r` becomes `http:%2F%2Ffoo%2Fb%20r`

fangfufu commented:
> So originally my thinking was to have a function that behaves like the libcurl function, but allows you specify what you want to escape.

No, in this situation you should not be writing code that duplicate what libcurl does. 

You can remove the protocol part of the URL, please refer to ``make_link_relative`` in ``link.c``

Have a look at https://curl.se/libcurl/c/curl_easy_getinfo.html

I am sorry but I don't like the current implementation, as it is a hack for the space character.

fangfufu commented:
Just for the record, this is for fixing https://github.com/fangfufu/httpdirfs/issues/138

fangfufu commented:
This pull request will not be merged, as you said earlier, it is basically a hack. If you are still interested in working on it, please submit something that actually addresses issue https://github.com/fangfufu/httpdirfs/issues/138. This problem actually has occurred in https://github.com/fangfufu/httpdirfs/issues/127 and https://github.com/fangfufu/httpdirfs/pull/137.

fangfufu commented:
There is also ``path_to_Link_recursive`` ( https://github.com/fangfufu/httpdirfs/blob/master/src/link.c#L827) which consumes each part of the URL to construct directories.

EyitopeIO commented:
I see. 
Okay, I'd do it correctly this time, not hacking for just the space character.

EyitopeIO commented:
Okay, I'd try one more time but with the understanding that the fix should generally address characters that could break URLs (e.g. in the related issues), and not just a hack for the space character.
